slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/AttestationImporter.java

```diff
@@ -0,0 +1,181 @@
+            LOG.debug("{} - already exists in database, not imported", attestationIdentifierString);
+          } else {
+            persist(jsonAttestation);
+          }
+        }
+      }
+    }
+    persistAttestationWatermark(handle, validator, minSourceTracker, minTargetTracker);
+  }
+
+  private void persist(final SignedAttestation jsonAttestation) {
+    signedAttestationsDao.insertAttestation(
+        handle,
+        new tech.pegasys.web3signer.slashingprotection.dao.SignedAttestation(
+            validator.getId(),
+            jsonAttestation.getSourceEpoch(),
+            jsonAttestation.getTargetEpoch(),
+            jsonAttestation.getSigningRoot()));
+    minSourceTracker.trackValue(jsonAttestation.getSourceEpoch());
+    minTargetTracker.trackValue(jsonAttestation.getTargetEpoch());
+  }
+
+  private void persistAttestationWatermark(
+      final Handle handle,
+      final Validator validator,
+      final OptionalMinValueTracker minSourceTracker,
+      final OptionalMinValueTracker minTargetTracker) {
+    final Optional<SigningWatermark> existingWatermark =
+        lowWatermarkDao.findLowWatermarkForValidator(handle, validator.getId());
+
+    final Optional<UInt64> newSourceWatermark =
+        findBestEpochWatermark(
+            minSourceTracker,
+            existingWatermark.flatMap(
+                watermark -> Optional.ofNullable(watermark.getSourceEpoch())));
+    final Optional<UInt64> newTargetWatermark =
+        findBestEpochWatermark(
+            minTargetTracker,
+            existingWatermark.flatMap(
+                watermark -> Optional.ofNullable(watermark.getTargetEpoch())));
+
+    if (newSourceWatermark.isPresent() && newTargetWatermark.isPresent()) {
+      LOG.info(
+          "Updating validator {} source epoch to {}",
+          validator.getPublicKey(),
+          newSourceWatermark.get());
+      LOG.info(
+          "Updating validator {} target epoch to {}",
+          validator.getPublicKey(),
+          newTargetWatermark.get());
+      lowWatermarkDao.updateEpochWatermarksFor(
+          handle, validator.getId(), newSourceWatermark.get(), newTargetWatermark.get());
+    } else if (newSourceWatermark.isPresent() != newTargetWatermark.isPresent()) {
+      throw new RuntimeException(
+          "Inconsistent data - no existing attestation watermark, "
+              + "and import only sets one epoch");
+    }
+  }
+
+  private Optional<UInt64> findBestEpochWatermark(
+      final OptionalMinValueTracker importedMin, final Optional<UInt64> currentWatermark) {
+    if (importedMin.compareTrackedValueTo(currentWatermark) > 0) {
+      return importedMin.getTrackedMinValue();
+    } else {
+      return currentWatermark;
+    }
+  }
+
+  private boolean nullAttestationAlreadyExistsInTargetEpoch(final UInt64 targetEpoch) {
+    return handle
+        .createQuery(
+            "SELECT validator_id, source_epoch, target_epoch, signing_root "
```

Comment:
nit: instead of selecting so many fields, you could've have use `select count(PK or *)...`, for instance. Since you only need to check whether record exists or not.